### DISCUSSION TITLE
How do I require whitespace?

### DISCUSSION THREAD

**Discussion by nathan-at-least** [zesterer/chumsky#126] [Q&A] [answered]
Hi, new to `chumsky` but I've almost replaced a half-written homebrew parser for a simple (non-recursive) grammar in less than an hour! Really slick experience, so thank you!

I want to require whitespace between tokens, but `chumsky::text::whitespace` allows zero instances.

The example in [`chumsky::primitive::filter`](https://docs.rs/chumsky/0.8.0/chumsky/primitive/fn.filter.html#examples) helped me write my own recipe that looks like this:

```
    let label = ident()
        .delimited_by(just('['), just(']'))
        .then(
            filter::<_, _, Simple<char>>(|c: &char| c.is_whitespace())
                .repeated()
                .at_least(1)
                .ignored(),
        )   
        .map(|(label, ())| label);
```

Is there a simpler approach I'm missing?

Would it be feasible and ergonomic to have the return type of `whitespace` behave similar to `repeated`, with modifier methods such as `whitespace().at_least(1)`?

**Comment by numToStr** [comment, marked as answer]
How about this?
```rust
let label = ident::<_, Simple<char>>()
    .delimited_by(just('['), just(']'))
    .then_ignore(just(' ').repeated().at_least(1));
```

**Reply by numToStr** [reply]
or (similar to your implementation)

```rust
let label = ident::<_, Simple<char>>()
    .delimited_by(just('['), just(']'))
    .then_ignore(filter(|c: &char| c.is_whitespace()).repeated().at_least(1));
```

**Reply by nathan-at-least** [reply]
This works! Thanks. Two bits I learned:

- Moving the type hint to `ident` is more concise.
- `.then_ignore` helps me avoid the custom `.map` to unwrap tuples.

Thanks!

**Comment by zesterer** [comment]
Hi! For now, that is the easiest setup. That said, `whitespace` internally uses `Repeated` so it should be possibly to simply reveal that fact to the caller, allowing it to invoke the usual methods. I'll open an issue.

**Reply by nathan-at-least** [reply]
I assume 1 or more whitespace is a common rule in many syntaxes, so it might be worth it. Thanks again, for a great tool.
